firebase-firestore/src/test/java/com/google/firebase/firestore/local/IndexFreeQueryEngineTest.java

```diff
@@ -0,0 +1,281 @@
+import static java.util.Arrays.asList;
+import static org.junit.Assert.assertEquals;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyInt;
+import static org.mockito.Mockito.doAnswer;
+
+import com.google.firebase.database.collection.ImmutableSortedMap;
+import com.google.firebase.database.collection.ImmutableSortedSet;
+import com.google.firebase.firestore.core.Query;
+import com.google.firebase.firestore.model.Document;
+import com.google.firebase.firestore.model.DocumentCollections;
+import com.google.firebase.firestore.model.DocumentKey;
+import com.google.firebase.firestore.model.MaybeDocument;
+import com.google.firebase.firestore.model.SnapshotVersion;
+import com.google.protobuf.ByteString;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.Config;
+
+@RunWith(RobolectricTestRunner.class)
+@Config(manifest = Config.NONE)
+public class IndexFreeQueryEngineTest {
+
+  private static final int TEST_TARGET_ID = 1;
+
+  private static final Document MATCHING_DOC_A =
+      doc("coll/a", 1, map("matches", true, "order", 1), Document.DocumentState.SYNCED);
+  private static final Document NON_MATCHING_DOC_A =
+      doc("coll/a", 1, map("matches", false, "order", 1), Document.DocumentState.SYNCED);
+  private static final Document PENDING_MATCHING_DOC_A =
+      doc("coll/a", 1, map("matches", true, "order", 1), Document.DocumentState.LOCAL_MUTATIONS);
+  private static final Document UDPATED_DOC_A =
+      doc("coll/a", 11, map("matches", true, "order", 1), Document.DocumentState.SYNCED);
+  private static final Document MATCHING_DOC_B =
+      doc("coll/b", 1, map("matches", true, "order", 2), Document.DocumentState.SYNCED);
+  private static final Document NON_MATCHING_DOC_B =
+      doc("coll/b", 1, map("matches", false, "order", 2), Document.DocumentState.SYNCED);
+  private static final Document UPDATED_MATCHING_DOC_B =
+      doc("coll/b", 11, map("matches", true, "order", 2), Document.DocumentState.SYNCED);
+
+  @Mock private LocalDocumentsView localDocumentsView;
+  @Mock private QueryCache queryCache;
+  private QueryEngine queryEngine;
+
+  private final Map<DocumentKey, Document> existingQueryResults = new HashMap<>();
+  private final Map<DocumentKey, Document> updatedQueryResults = new HashMap<>();
+  private boolean expectIndexFreeExecution;
+
+  @Before
+  public void setUp() {
+    MockitoAnnotations.initMocks(this);
+
+    expectIndexFreeExecution = false;
+    existingQueryResults.clear();
+    updatedQueryResults.clear();
+
+    queryEngine = new IndexFreeQueryEngine();
+    queryEngine.setLocalDocumentsView(localDocumentsView);
+
+    doAnswer(
```

Comment:
It seems like this could be done much more straightforwardly by just using a regular MemoryQueryCache. 

This would make `addExistingResult` call `queryCache.addMatchingKeys(doc.getKey(), TEST_TARGET_ID)` and avoid all these mocks, which don't port easily and end up being very brittle and hard to change.